modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +171,110 @@ private void ensureOpen() {
+        if (attempt == 1) {
+            logger.log(
+                level,
+                () -> format(
+                    "failed %s [%s/%s] at offset [%s] with purpose [%s]",
+                    action,
+                    blobStore.bucket(),
+                    blobKey,
+                    start + currentOffset,
+                    purpose.getKey()
+                ),
+                e
+            );
+            return;
+        }
 
         final long meaningfulProgressSize = Math.max(1L, blobStore.bufferSizeInBytes() / 100L);
         final long currentStreamProgress = Math.subtractExact(Math.addExact(start, currentOffset), currentStreamFirstOffset);
         if (currentStreamProgress >= meaningfulProgressSize) {
             failuresAfterMeaningfulProgress += 1;
         }
-        final Supplier<String> messageSupplier = () -> format(
-            """
-                failed reading [%s/%s] at offset [%s]; this was attempt [%s] to read this blob which yielded [%s] bytes; in total \
-                [%s] of the attempts to read this blob have made meaningful progress and do not count towards the maximum number of \
-                retries; the maximum number of read attempts which do not make meaningful progress is [%s]""",
-            blobStore.bucket(),
-            blobKey,
-            start + currentOffset,
-            attempt,
-            currentStreamProgress,
-            failuresAfterMeaningfulProgress,
-            maxAttempts
+        logger.log(
+            level,
+            () -> format(
+                """
+                    failed %s [%s/%s] at offset [%s] with purpose [%s]; \
+                    this was attempt [%s] to read this blob which yielded [%s] bytes; in total \
+                    [%s] of the attempts to read this blob have made meaningful progress and do not count towards the maximum number of \
+                    retries; the maximum number of read attempts which do not make meaningful progress is [%s]""",
+                action,
+                blobStore.bucket(),
+                blobKey,
+                start + currentOffset,
+                purpose.getKey(),
+                attempt - 1,
+                currentStreamProgress,
+                failuresAfterMeaningfulProgress,
+                maxRetriesForNoMeaningfulProgress()
+            ),
+            e
         );
-        if (attempt >= maxAttempts + failuresAfterMeaningfulProgress) {
-            final var finalException = addSuppressedExceptions(e);
-            logger.warn(messageSupplier, finalException);
-            throw finalException;
+    }
+
+    private boolean shouldRetry() {
+        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
+            return false;
         }
-        logger.debug(messageSupplier, e);
-        attempt += 1;
-        if (failures.size() < MAX_SUPPRESSED_EXCEPTIONS) {
-            failures.add(e);
+        if (purpose == OperationPurpose.INDICES) {
```

Comment:
Yeah I think it can be a follow-up. Existing issues that I have studied so far are all for indices data. Failure to read translog data may have a different failure pattern. I can look into them after this.